src/core/src/service_interfaces/TelemetryWriter.py

```diff
@@ -174,25 +181,41 @@ def __delete_older_events(self):
         if self.__get_events_dir_size() >= Constants.TELEMETRY_DIR_SIZE_LIMIT_IN_BYTES:
             raise Exception("Older event files were not deleted. Current event will not be sent to telemetry as events directory size exceeds maximum limit")
 
-    def write_event_using_temp_file(self, folder_path, data, mode='w'):
-        """ Writes to a temp file in a single operation and then moves/overrides the original file with the temp """
+    def get_file_and_content_to_write(self, folder_path, data):
+        """ Identifies the file where the event is to be written can be an existing event file or a new one depending upon the size restrictions. If event is to be written to an existing file, fetches retains it's content """
         file_path = self.__get_event_file_path(folder_path)
-        prev_events = []
+        all_events = []
+        if os.path.exists(file_path):
+            file_size = self.get_file_size(file_path)
+            # if file_size exceeds max limit, sleep for 1 second, so the event can be written to a new file since the event file name is a timestamp
+            if file_size >= Constants.TELEMETRY_EVENT_FILE_SIZE_LIMIT_IN_BYTES:
+                time.sleep(1)
+                file_path = self.__get_event_file_path(folder_path)
+            else:
+                all_events = self.__fetch_events_from_previous_file(file_path)  # fetches existing content within the file
+        all_events.append(data)
+        return file_path, all_events
+
+    def ensure_event_file_max_count_is_met(self):
+        """ Ensures the # of event files that can be written per time unit restriction is met. Returns False if the any updates are required after the restriction enforcement. For eg: file_name is a timestamp and should be modified if a wait is added here """
+        if (datetime.datetime.utcnow() - self.start_time_for_event_file_throttle_check).total_seconds() < Constants.TELEMETRY_MAX_TIME_FOR_EVENT_FILE_THROTTLE:
+            # If event file count limit reached before time period, wait out the remaining time
+            if self.event_file_count >= Constants.TELEMETRY_MAX_EVENT_FILE_THROTTLE_COUNT:
+                time.sleep((datetime.datetime.utcnow() - self.start_time_for_event_file_throttle_check).total_seconds())
```

Comment:
Calculate the time to sleep, verify again it's not more than the upper limit of time to sleep, log it and then sleep. Gaps in execution time must be ascertainable in logs and telemetry.